Use GtkSourceView for undo/redo functionality

takluyver opened the pull request:
(Hello from EuroPython :-)

This isn't quite ready to go, but the day is drawing to a close, so I wanted to open a PR with what I'd got done. The aim is to use [GtkSourceView](https://wiki.gnome.org/Projects/GtkSourceView) for the day text editor, which brings built-in undo history and syntax highlighting. This could save a fair bit of code in RedNotebook, and also fix some UX quirks with undo.

I store the most recent 10 'source buffers' used, which preserve undo history and cursor position for the last 10 days viewed. Once a buffer falls out of this recent list, the undo history and cursor position for it are lost. It would be easy to increase the size of the recent list. I started small to make it easy to check the behaviour.

Bits to figure out:

* The `UndoRedoManager` in rednotebook also stores actions for the categories view. This [can no longer be enabled](2ef88a7c75e7cd53ea6c35c99f4652b9a90b14f6), but the code is still there. If the plan is to remove this entirely, it's easy, because we can just use the GtkSourceView undo system (as in this PR). If the categories view is coming back in some form and undo should remain integrated, it's more complex.
* The `txt2tags` highlighting mode for GtkSourceView doesn't seem to be very good. This is defined in [an XML file](https://gitlab.gnome.org/GNOME/gtksourceview/blob/master/data/language-specs/t2t.lang). I think it's possible to ship a modified XML file separately and use it if necessary, but we're also interested in the possibility of using markdown, since it's so much more widely known.

jendrikseipp commented:
Nice patch! Switching to GtkSourceView is a great idea and I'm glad you started working on this.

Removing CategoriesTreeView is a longterm goal, but it requires some more preparation. However, we can already remove undo/redo functionality for CategoriesTreeView now. This should allow removing most of the code that is now reported as dead and the code that is currently commented out.

I think it would be good to ship our own `t2t.lang` XML file. The official version seems to be very limited in functionality. It would be great if you could add support for the most important constructs like images, named links, plain URLs and strike-through. Maybe the `markdown.lang` can serve as an example for some of these. It would be nice to switch to Markdown right away, but before we can make the switch, we need a good idea for how to migrate users' journals...

takluyver commented:
Thanks! Following your suggestions, I have:

- Removed undo/redo from CategoriesTreeView.
- Removed UndoRedoManager as there was no code left using it
- Added a copy of `t2t.lang` and told it to use this

At the moment,  `t2t.lang` is unmodified from the gtksourceview version, so there's no difference in the syntax highlighting. The next step is figuring out how the language definition files work and improving it.

jendrikseipp commented:
Sounds good!

takluyver commented:
Before changes:

![screenshot from 2018-08-01 15-54-51](https://user-images.githubusercontent.com/327925/43533123-f5e8f578-95ab-11e8-8478-172cde515a52.png)

After changes:

![screenshot from 2018-08-01 15-54-34](https://user-images.githubusercontent.com/327925/43533130-f9dad5c0-95ab-11e8-97c9-c439b190885b.png)

I've got a ~2 hours flight coming up - I've opened tabs for the txt2tags syntax and the [GtkSourceView highlighting definition format](https://developer.gnome.org/gtksourceview/stable/syntax-highlighting-reference.html), so I might spend some more time expanding this.

takluyver commented:
I've now added a custom style as well as the custom language definition. This reduces its ability to blend in to system style preferences, but it lets us define styles better suited to a markup language like txt2tags. GtkSourceView's default style classes are defined for a programming language (naturally enough): constants, identifiers, types, and so on. You can kind of shoehorn bits of a markup language into the classes to make the highlighter show when things match, but you can do a much better job by defining some new classes.

I want to add some highlighting around links, then I think this is pretty much ready.

takluyver commented:
OK, I'm reasonably happy with this. The link highlighting is rather simplified relative to the real rules, because the actual txt2tags implementation assembles a complex regex from various pieces, and I didn't try to reproduce it fully.

Here's what the highlighting looks like now

![screenshot from 2018-08-02 11-54-59](https://user-images.githubusercontent.com/327925/43577148-5db4c5c0-964b-11e8-9890-5584f75e5ac9.png)

jendrikseipp commented on rednotebook/gui/main_window.py:
```diff
@@ -663,48 +668,80 @@ def show_message(self, title, msg, msg_type):
         else:
             self.statusbar.show_message(title, msg, msg_type)
 
+    def update_undo_redo_buttons(self, _gobj=None, force_disable=False):
```
The `_gobj` kwarg seems to be unused.

takluyver replied:
It is, but this method is connected with a GTK signal handler, and it seems that it passes the GObject that sent the signal as the first argument. So without it, `force_disable` is given some unwanted object. I haven't read documentation about this, it was just trial and error, so maybe you know of a better way to deal with that.

takluyver replied:
Done, but I've left `_gobject=None` so it's an optional argument - the method is also called from Python code (e.g. when switching from the HTML preview back to edit mode), and it doesn't seem right to pass a useless parameter in.

jendrikseipp replied:
Sounds good to me.

jendrikseipp commented on rednotebook/files/rednotebook-highlight-style.xml:
```diff
@@ -0,0 +1,136 @@
+ This file is part of GtkSourceView
+
+ Copyright (C) 2006-2007 GtkSourceView team
+ Author: Michael Monreal <michael.monreal@gmail.com>
```
Could you please extend/update the copyright and authors information as you see fit in the two new XML files? Same for the `<author>` field below. Also, maybe we should add a comment explaining that the files come from GtkSourceView but were adapted for RedNotebook.

jendrikseipp commented on rednotebook/gui/categories.py:
```diff
@@ -25,15 +25,14 @@
 
         self.main_window = main_window
-        self.undo_redo_manager = main_window.undo_redo_manager
+        # self.undo_redo_manager = main_window.undo_redo_manager
```
This line can be removed.

jendrikseipp commented on rednotebook/gui/main_window.py:
```diff
@@ -663,48 +668,80 @@ def show_message(self, title, msg, msg_type):
+        buf.set_text(day.text)
+        buf.end_not_undoable_action()
+
+        while len(self.recent_buffers) > self.n_recent_buffers:
```
Shouldn't an `if` instead of `while` suffice since only one buffer is added at a time?

takluyver replied:
It should suffice; I was doing this to be defensive. If something else adds further buffers, this will always bring the length back to the limit defined as `n_recent_buffers`. I don't mind changing it if you prefer, though.

takluyver replied:
OK, done.

jendrikseipp commented:
Really nice work on the highlighting! Some things even work better than before, like multiline verbatim text.

jendrikseipp commented:
Thanks for making the changes! I'll test the patch locally now.

jendrikseipp commented:
Sorry for adding commits from

takluyver commented:
I looked into the [style definitions](https://developer.gnome.org/gtksourceview/stable/style-reference.html) a bit before - I don't think it's possible to select a monospace font for monospace blocks, unfortunately.

The bits with the links should be possible, but I'll need to figure out what the rules are for matching them.

takluyver commented:
I added highlighting for bare links and URLs:

![screenshot from 2018-08-10 22-24-05](https://user-images.githubusercontent.com/327925/43979830-4c2e5350-9cec-11e8-8be6-470ef6c86711.png)

I can't find the quotation marks around URLs either in the [online documentation for txt2tags](https://txt2tags.org/doc/english/rules#link) or in the regexes in the source code to do the conversion. Are they something that rednotebook adds on top of t2t? :confused:

jendrikseipp commented:
The double quotation marks are needed to escape whitespace in links. This way of escaping is common in txt2tags. You can find the regexes at the top of util/markup.py (`REGEX_PIC`, `REGEX_NAMED_LINK`).

takluyver commented:
Aha, thanks. I copied those into the language definition, and here we are:

![screenshot from 2018-08-11 09-09-47](https://user-images.githubusercontent.com/327925/43989253-b41f28a4-9d46-11e8-93c3-81d40ce2d673.png)

takluyver commented:
I do see the same warnings. It's weird that the persist even if you set the style before the language definition. Maybe the style is loaded asynchronously somehow? It can't seem any problematic effects of this - the styles still seem to be applied correctly.

jendrikseipp commented:
I was able to fix the warnings after finding some hints here: https://ubuntuforums.org/showthread.php?t=1150569&s=d0a921a1217b7f56d7550265b08376cc&p=10316801#post10316801

I still have to look into making this work on Windows.

jendrikseipp commented:
While testing I noticed that some things are not working anymore. Could you look into the following, please?

* templates: they should probably have a temporary undo/redo buffer that is only used while the template is being edited and discarded afterwards.
* search: the searched word should be highlighted in the day text view.

takluyver commented:
Have a go now.

While figuring out the highlighting thing, I also noticed that `Editor.scroll_to_text()` had a hack to do a kind-of case insensitive search. It appears that Gtk now has a [flag for case insensitive search](https://developer.gnome.org/gtk3/stable/GtkTextIter.html#GtkTextSearchFlags), so I dropped the workaround. I don't know what minimum version of Gtk that needs, but it's working for me.

takluyver commented:
The CASE_INSENSITIVE flag appears to have been in GTK for many years, so I think it's safe to rely on it. https://gitlab.gnome.org/GNOME/gtk/commit/794e0446e94f8d0f994cda4eae47c89370e3141c

jendrikseipp commented:
Thanks! I'll have a look as soon as time permits.

jendrikseipp commented:
Code looks perfect. It's very nice to see that you simplified so many rather involved parts of the code. Thanks for your help!

takluyver commented:
Thanks! I'm glad something useful came out of my sprints time. :-)
